src/platform/backends/lxd/lxd_vm_image_vault.cpp

```diff
@@ -225,13 +195,64 @@ bool mp::LXDVMImageVault::has_record_for(const std::string& name)
 
 void mp::LXDVMImageVault::prune_expired_images()
 {
-    mpl::log(mpl::Level::trace, category, "Pruning expired images not implemented");
+    auto json_reply = lxd_request(manager, "GET", QUrl(QString("%1/images").arg(base_url.toString())));
+
+    auto images = json_reply["metadata"].toArray();
+
+    for (const auto image : images)
+    {
+        auto image_info = image.toObject();
+        auto last_used_count =
+            QDateTime::fromString(image_info["last_used_at"].toString(), Qt::ISODateWithMs).toMSecsSinceEpoch();
+        auto last_used = std::chrono::system_clock::time_point(std::chrono::milliseconds(last_used_count));
+
+        if (image_info.contains("update_source") && last_used + days_to_expire <= std::chrono::system_clock::now())
```

Comment:
Yes, it's LXD's way of storing image info when the image was downloaded via an alias vs. a hash (or fingerprint) and indicates to us that it's an alias type image and we can update it.

Reply:
Sorry, I thought this was the `update_images()` function.  So what I am trying to accomplish is that if an image was launched via a hash, then it is not pruned no matter what.  I believe the default is that hash based images are never expired.

Reply:
I don't believe that was the desire… We never update images from hash queries (because… how?), but we prune them just the same? Is that not the case in the default image host?

Reply:
The `Query` class has a `bool persistent` field that I believe was intended to signify that an image was launched based on hash.  `DefaultVMImageVault::prune_expired_images()` does check if persistent is set when deciding if it should delete an image.  That said, it looks like the daemon _always_ sets `persistent` to `false`, so the field is really a throw away.  I wonder if that is a bug or if we just lost sense of that that part of the query is for???

Reply:
Ok, I went back in time and looked at a diagram you made about the queries, etc. and this is what you said about `persistent`:
"Persistent queries can only be created via GUI, by checking a box next to "Latest LTS", "16.04 LTS", "14.04 LTS", "17.10", "Devel" etc."

So I'm wrong and will fix this :smiley:

Reply:
:win:

Reply:
Fixed